espnet/nets/pytorch_backend/e2e_asr_mix_transformer.py

```diff
@@ -0,0 +1,452 @@
+            for i in range(self.num_spkrs):
+                (
+                    pred_pad[i],
+                    pred_mask[i],
+                    loss_att[i],
+                    acc[i],
+                ) = self.decoder_and_attention(
+                    hs_pad[i], hs_mask[i], ys_pad[i], batch_size
+                )
+
+            # 4. compute attention loss
+            # The following is just an approximation
+            loss_att = sum(map(lambda x: x[0] * x[1], zip(loss_att, ys_out_len))) / sum(
+                ys_out_len
+            )
+            self.acc = sum(map(lambda x: x[0] * x[1], zip(acc, ys_out_len))) / sum(
+                ys_out_len
+            )
+
+            # 5. compute cer/wer
+            if self.training or self.error_calculator is None:
+                cer, wer = None, None
+            else:
+                ys_hat = pred_pad.argmax(dim=-1)
+                cer, wer = self.error_calculator(ys_hat.cpu(), ys_pad.cpu())
+
+        # copyied from e2e_asr
+        alpha = self.mtlalpha
+        if alpha == 0:
+            self.loss = loss_att
+            loss_att_data = float(loss_att)
+            loss_ctc_data = None
+        elif alpha == 1:
+            self.loss = loss_ctc
+            loss_att_data = None
+            loss_ctc_data = float(loss_ctc)
+        else:
+            self.loss = alpha * loss_ctc + (1 - alpha) * loss_att
+            loss_att_data = float(loss_att)
+            loss_ctc_data = float(loss_ctc)
+
+        loss_data = float(self.loss)
+        if loss_data < CTC_LOSS_THRESHOLD and not math.isnan(loss_data):
+            self.reporter.report(
+                loss_ctc_data, loss_att_data, self.acc, cer_ctc, cer, wer, loss_data
+            )
+        else:
+            logging.warning("loss (=%f) is not correct", loss_data)
+        return self.loss
+
+    def decoder_and_attention(self, hs_pad, hs_mask, ys_pad, batch_size):
+        """Forward decoder and attention loss."""
+        # forward decoder
+        ys_in_pad, ys_out_pad = add_sos_eos(ys_pad, self.sos, self.eos, self.ignore_id)
+        ys_mask = target_mask(ys_in_pad, self.ignore_id)
+        pred_pad, pred_mask = self.decoder(ys_in_pad, ys_mask, hs_pad, hs_mask)
+
+        # compute attention loss
+        loss_att = self.criterion(pred_pad, ys_out_pad)
+        acc = th_accuracy(
+            pred_pad.view(-1, self.odim), ys_out_pad, ignore_label=self.ignore_id
+        )
+        return pred_pad, pred_mask, loss_att, acc
+
+    def encode(self, x):
+        """Encode acoustic features.
+
+        :param ndarray x: source acoustic feature (T, D)
+        :return: encoder outputs
+        :rtype: torch.Tensor
+        """
+        self.eval()
+        x = torch.as_tensor(x).unsqueeze(0)
+        enc_output, _ = self.encoder(x, None)
+        return enc_output
+
+    def recog(self, enc_output, recog_args, char_list=None, rnnlm=None, use_jit=False):
```

Comment:
One future work (not now) is to implement this beam search under api_v2. 
https://github.com/espnet/espnet/issues/1384
Please check it.